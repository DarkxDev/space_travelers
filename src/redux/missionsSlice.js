import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const fetchMissions = createAsyncThunk(
  'missions/fetchMissions',
  async () => {
    const response = await axios.get('https://api.spacexdata.com/v3/missions');
    return response;
  },
);

const missionsSlice = createSlice({
  name: 'missions',
  initialState: {
    missions: [],
    status: 'idle',
    error: null,
  },
  reducers: {
    reserveMission: (state, action) => {
      const missionId = action.payload;
      const updatedMissions = state.missions.map((mission) => (mission.mission_id === missionId
        ? { ...mission, reserved: true }
        : mission));
      return {
        ...state,
        missions: updatedMissions,
      };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchMissions.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(fetchMissions.fulfilled, (state, action) => {
        const missions = action.payload.data.map((mission) => ({
          mission_id: mission.mission_id,
          mission_name: mission.mission_name,
          description: mission.description,
        }));
        return {
          ...state,
          status: 'succeeded',
          missions,
        };
      })
      .addCase(fetchMissions.rejected, (state, action) => ({
        ...state,
        status: 'failed',
        missions: action.error.message,
      }));
  },
});

export { missionsSlice, fetchMissions };
export const { reserveMission } = missionsSlice.actions;
