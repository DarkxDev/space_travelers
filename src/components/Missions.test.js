import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';
import MissionsPage from './Missions';
import { fetchMissions, leaveMission, reserveMission } from '../redux/missions/missionsSlice';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('../redux/missions/missionsSlice', () => ({
  fetchMissions: jest.fn(),
  leaveMission: jest.fn(),
  reserveMission: jest.fn(),
}));

describe('Missions Page', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn());
    useSelector.mockImplementation((selectorFn) => selectorFn({
      missions: {
        missions: [
          {
            mission_id: '1',
            mission_name: 'Mission 1',
            description: 'Mission 1 description',
            reserved: false,
          },
          {
            mission_id: '2',
            mission_name: 'Mission 2',
            description: 'Mission 2 description',
            reserved: true,
          },
        ],
      },
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('component matches snapshot', () => {
    const { asFragment } = render(<MissionsPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('dispatches fetchMissions action if missions array is empty', () => {
    useSelector.mockImplementation((selectorFn) => selectorFn({
      missions: {
        missions: [],
      },
    }));

    render(<MissionsPage />);

    expect(fetchMissions).toHaveBeenCalledTimes(1);
  });

  it('dispatches reserveMission when Join Mission button is clicked', () => {
    render(<MissionsPage />);

    const joinMissionButton = screen.getByText('Join Mission');
    fireEvent.click(joinMissionButton);

    expect(reserveMission).toHaveBeenCalledTimes(1);
    expect(reserveMission).toHaveBeenCalledWith('1'); // assuming mission_id is '1'
  });

  it('dispatches leaveMission when Leave Mission button is clicked', () => {
    render(<MissionsPage />);

    const leaveMissionButton = screen.getByText('Leave Mission');
    fireEvent.click(leaveMissionButton);

    expect(leaveMission).toHaveBeenCalledTimes(1);
    expect(leaveMission).toHaveBeenCalledWith('2'); // assuming mission_id is '2'
  });
});
