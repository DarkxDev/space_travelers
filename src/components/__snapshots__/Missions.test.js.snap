// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Page Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="missions-section"
  >
    <hr
      class="missions-divider"
    />
    <table>
      <thead>
        <tr>
          <th
            class="table-head"
          >
            Mission
          </th>
          <th
            class="table-head"
          >
            Description
          </th>
          <th
            class="table-head"
          >
            Status
          </th>
          <th
            class="table-head"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="mission-name"
          >
            Mission 1
          </td>
          <td>
            <p
              class="mission-description"
            >
              Mission 1 description
            </p>
          </td>
          <td>
            <p
              class="status-badge status-badge-inactive"
            >
              NOT A MEMBER
            </p>
          </td>
          <td>
            <button
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
        <tr>
          <td
            class="mission-name"
          >
            Mission 2
          </td>
          <td>
            <p
              class="mission-description"
            >
              Mission 2 description
            </p>
          </td>
          <td>
            <p
              class="status-badge status-badge-active"
            >
              Active Member
            </p>
          </td>
          <td>
            <button
              type="button"
            >
              Leave Mission
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`Missions Page component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="missions-section"
  >
    <hr
      class="missions-divider"
    />
    <table>
      <thead>
        <tr>
          <th
            class="table-head"
          >
            Mission
          </th>
          <th
            class="table-head"
          >
            Description
          </th>
          <th
            class="table-head"
          >
            Status
          </th>
          <th
            class="table-head"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="mission-name"
          >
            Mission 1
          </td>
          <td>
            <p
              class="mission-description"
            >
              Mission 1 description
            </p>
          </td>
          <td>
            <p
              class="status-badge status-badge-inactive"
            >
              NOT A MEMBER
            </p>
          </td>
          <td>
            <button
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
        <tr>
          <td
            class="mission-name"
          >
            Mission 2
          </td>
          <td>
            <p
              class="mission-description"
            >
              Mission 2 description
            </p>
          </td>
          <td>
            <p
              class="status-badge status-badge-active"
            >
              Active Member
            </p>
          </td>
          <td>
            <button
              type="button"
            >
              Leave Mission
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`Missions Page should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="missions-section"
  >
    <hr
      class="missions-divider"
    />
    <table>
      <thead>
        <tr>
          <th
            class="table-head"
          >
            Mission
          </th>
          <th
            class="table-head"
          >
            Description
          </th>
          <th
            class="table-head"
          >
            Status
          </th>
          <th
            class="table-head"
          >
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="mission-name"
          >
            Mission 1
          </td>
          <td>
            <p
              class="mission-description"
            >
              Mission 1 description
            </p>
          </td>
          <td>
            <p
              class="status-badge status-badge-inactive"
            >
              NOT A MEMBER
            </p>
          </td>
          <td>
            <button
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
        <tr>
          <td
            class="mission-name"
          >
            Mission 2
          </td>
          <td>
            <p
              class="mission-description"
            >
              Mission 2 description
            </p>
          </td>
          <td>
            <p
              class="status-badge status-badge-active"
            >
              Active Member
            </p>
          </td>
          <td>
            <button
              type="button"
            >
              Leave Mission
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
